File: 94-binary-tree-inorder-traversal/94-binary-tree-inorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def getRightMostNode(self, root: TreeNode, curr : TreeNode) -> TreeNode:
        while root.right != None and root.right != curr:
            root = root.right
        return root;
    
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        ans = []
        curr = root
        while curr != None:
            if curr.left == None:
                ans.append(curr.val)
                curr = curr.right
            else:
                rightMostNode = self.getRightMostNode(curr.left, curr)
                if rightMostNode.right == None: # creating the thread
                    rightMostNode.right = curr
                    curr = curr.left
                else: # breaking the thread
                    ans.append(curr.val)
                    curr = curr.right
                    rightMostNode.right = None
        
        return ans

File: 94-binary-tree-inorder-traversal/94-binary-tree-inorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    void go(TreeNode* root, vector<int> &v) {
        if(root == NULL) return;
        go(root->left, v);
        v.push_back(root->val);
        go(root->right, v);
    }
public:
    vector<int> inorderTraversal(TreeNode* root) {
        vector<int> v;
        go(root, v);
        return v;
    }
};